internal/topology/variables/patch_validation.go

```diff
@@ -0,0 +1,222 @@
+					path.Index(i),
+					jsonPatch,
+					"jsonPatch path must start with \"/spec/\".",
+				))
+		}
+
+		if _, err := jsonptr.Parse(jsonPatch.Path); err != nil {
+			allErrs = append(allErrs,
+				field.Invalid(
+					path.Child("path").Index(i),
+					jsonPatch,
+					fmt.Sprintf("jsonPatch path must be a valid JSON pointer (RFC 6901). %s not valid.", jsonPatch.Path),
+				))
+		}
+
+		// return early if the jsonPatch does not have "replace" or "add" op. Additional validation is not needed.
+		if jsonPatch.Op != "add" && jsonPatch.Op != "replace" {
+			return allErrs
+		}
+
+		if jsonPatch.Value == nil && jsonPatch.ValueFrom == nil {
+			allErrs = append(allErrs,
+				field.Invalid(
+					path.Index(i),
+					jsonPatch,
+					"jsonPatch must define one of Value or ValueFrom.",
+				))
+		}
+		if jsonPatch.Value != nil && jsonPatch.ValueFrom != nil {
+			allErrs = append(allErrs,
+				field.Invalid(
+					path.Index(i),
+					jsonPatch,
+					"jsonPatch can not define both Value and ValueFrom.",
+				))
+		}
+		if jsonPatch.ValueFrom != nil && jsonPatch.ValueFrom.Template == nil && jsonPatch.ValueFrom.Variable == nil {
+			allErrs = append(allErrs,
+				field.Invalid(
+					path.Index(i),
+					jsonPatch,
+					"ValueFrom must set either \"/Template or Variable/\".",
+				))
+		}
+		if jsonPatch.ValueFrom != nil && jsonPatch.ValueFrom.Template != nil && jsonPatch.ValueFrom.Variable != nil {
+			allErrs = append(allErrs,
+				field.Invalid(
+					path.Index(i),
+					jsonPatch,
+					"ValueFrom can not set both \"/Template and Variable/\".",
+				))
+		}
+
+		//TODO: return early if the jsonPatch has the "replace" op. Additional validation is not needed.
+		// if jsonPatch.Op == "replace" {
+		//	return allErrs
+		//}
+
+		// TODO: only the indexes `0` (prepend) and `-` (append) in combination with the `add` operation are allowed (append and prepend are the only allowed array modifications).
+		if strings.Contains(jsonPatch.Path, "ARRAY INDEXING") {
+			allErrs = append(allErrs,
+				field.Invalid(
+					path.Index(i),
+					jsonPatch,
+					fmt.Sprintf("Arrays can only be accessed using [0] (prepend) or [-] (append). %s not valid", "WHAT IS THIS"),
+				))
+		}
+		if jsonPatch.ValueFrom != nil && jsonPatch.ValueFrom.Template != nil {
+			// Error if template can not be parsed.
+			_, err := template.New("tpl").Parse(*jsonPatch.ValueFrom.Template)
+			if err != nil {
+				allErrs = append(allErrs,
+					field.Invalid(
+						path.Child("valueFrom", "template"),
+						jsonPatch.ValueFrom.Template,
+						fmt.Sprintf("Template can not be parsed. Error %s", err),
+					))
+			}
+		}
+
+		//TODO: `variable` (if set): should be a valid variable
+		variableSet := getClusterClassVariablesMap(variables)
+		// If set validate that the variable is valid.
+		if jsonPatch.ValueFrom != nil && jsonPatch.ValueFrom.Variable != nil {
+			// If the variable looks like a builtin it's valid.
+			// TODO: Should we try to check if this is a valid builtin?
```

Comment:
@sbueringer is there a simple way to get all the supported builtin variables (and keep this in sync with the internal struct we recently introduced)?

Reply:
Just talked with Killian about this, there are a few bad ways to do this:
* create constants for all builtin variables and expose them somewhere as valid builtin variables (duplicate info to the struct JSON tags)
* go with reflection over the builtin struct

Once we also have user-defined "complex" (object/array) variables it gets even worse. I'm not sure there is a way to implement a "full" validation without restructuring and exposing parts of our patch code (our idea was to defer this out of this PR)

Reply:
Currently the buitlin variables as they are implemented are not even part of our currently internal patch API. So we - the same as external patch implementers - have to way to consume how buitlin variables are structured and which exist.

Reply:
let's just define a list of well know built-in variables and comment on the internal types that any change should be synchronized with this list